parsec/parsec_mpi_funnelled.c

```diff
@@ -713,32 +805,26 @@ int
 mpi_no_thread_tag_unregister(parsec_ce_tag_t tag)
 {
     mpi_funnelled_tag_t *tag_struct = &parsec_mpi_funnelled_array_of_registered_tags[tag];
-    if(NULL == tag_struct->buf) {
+    if( (PARSEC_CE_TAG_STATUS_INACTIVE == tag_struct->status) ||
+        (PARSEC_CE_TAG_STATUS_DISABLE == tag_struct->status) ) {
         parsec_inform("Tag %ld is not registered\n", (int)tag);
-        return 0;
+        return PARSEC_SUCCESS;
     }
-
-    /* remove this tag from the arrays */
-    /* WARNING: Assumed after this no wait or test will be called on
-     * array_of_requests
-     */
-    int i, flag;
-    MPI_Status status;
-
-    for(i = tag_struct->start_idx; i < tag_struct->start_idx + EACH_STATIC_REQ_RANGE; i++) {
-#if !defined(CRAY_MPICH_VERSION)
-        // MPI Cancel broken on Cray
-        MPI_Cancel(&array_of_requests[i]);
-        MPI_Test(&array_of_requests[i], &flag, &status);
-#endif
-        MPI_Request_free(&array_of_requests[i]);
-        assert( MPI_REQUEST_NULL == array_of_requests[i] );
+    parsec_atomic_lock(&parsec_ce_am_build_lock);
+    if( PARSEC_CE_TAG_STATUS_ENABLE == tag_struct->status ) {
+        /* requests not yet create, change the status and return */
+        tag_struct->status = PARSEC_CE_TAG_STATUS_INACTIVE;
+        return PARSEC_SUCCESS;
     }
+    tag_struct->status = PARSEC_CE_TAG_STATUS_DISABLE;
+    /* if the engine is active, notify it to rebuild the arrays of requests at the next cycle */
+    parsec_ce_am_design_version++;
+    parsec_atomic_unlock(&parsec_ce_am_build_lock);
 
-    free(tag_struct->buf[0]);
-    free(tag_struct->buf);
+    /* TODO: This part is not yet safe, we need to know if the engine is up or not */
+    mpi_no_thread_tag_unregister_internal(tag);
```

Comment:
Is it not sufficient to defer this until the engine comes up? There won't be any AMs before and the first thing the progress thread does is to handle the internal registration.

Reply:
The real problem is that by allowing too much we are screwing ourselves. If we are using this is a very sensible way, then by the time the upper level unregister the tags the communication engine is already down, so there will be nobody to execute the tag unregister. But, I guess I can take care of this in the ce_fini.